x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
+                List<Map.Entry<String, Long>> nodeShardsStorageList = new ArrayList<>(nodeShardsStorageBytes.entrySet());
+                nodeShardsStorageList.sort((o1, o2) -> o2.getValue().compareTo(o1.getValue()));
+                Optional<String> nodeId = Optional.empty();
+                for (Map.Entry<String, Long> entry : nodeShardsStorageList) {
+                    // we prefer to select the node which contains the maximum shards storage bytes of the index from the valid node list
+                    if (validNodeIds.contains(entry.getKey())) {
+                        nodeId = Optional.of(entry.getKey());
+                        break;
+                    }
+                }
+
+                // if we cannot find a node which contains any shard of the index,
+                // shuffle the valid node list and select randomly
+                if (nodeId.isEmpty()) {
+                    List<String> list = new ArrayList<>(validNodeIds);
+                    Randomness.shuffle(list);
+                    nodeId = list.stream().findAny();
+                }
 
-            if (nodeId.isPresent()) {
-                Settings settings = Settings.builder()
+                if (nodeId.isPresent()) {
+                    Settings settings = Settings.builder()
                         .put(IndexMetadata.INDEX_ROUTING_REQUIRE_GROUP_SETTING.getKey() + "_id", nodeId.get()).build();
-                UpdateSettingsRequest updateSettingsRequest = new UpdateSettingsRequest(indexName)
+                    UpdateSettingsRequest updateSettingsRequest = new UpdateSettingsRequest(indexName)
                         .masterNodeTimeout(TimeValue.MAX_VALUE)
                         .settings(settings);
-                getClient().admin().indices().updateSettings(updateSettingsRequest,
+                    getClient().admin().indices().updateSettings(updateSettingsRequest,
                         ActionListener.wrap(response -> listener.onResponse(true), listener::onFailure));
-            } else {
-                // No nodes currently match the allocation rules, so report this as an error and we'll retry
-                logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
-                listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" + indexName + "] onto" +
-                    " prior to shrink"));
-            }
+                } else {
+                    // No nodes currently match the allocation rules or have no enough disk bytes,
+                    // so report this as an error and we'll retry
+                    logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
+                    listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" + indexName +
+                        "] onto prior to shrink"));
+                }
+            }, listener::onFailure));
```

Comment:
This requires some consideration, for example, what happens if the nodes stats times out, should we fail open (still try to find a node), or fail closed?

I think we should at least fill out the failure handler so that if the nodes stats call fails, the message in the exception that ILM explain will show is more human readable. Something like "failed to retrieve disk information to select a single node for primary shard allocation".